rust/kernel/src/sync.rs

```diff
@@ -0,0 +1,139 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use core::cell::UnsafeCell;
+use core::ops::{Deref, DerefMut};
+use core::fmt;
+
+use crate::bindings;
+
+/*#[cfg(CONFIG_DEBUG_SPINLOCK)]
+compile_error!("CONFIG_DEBUG_SPINLOCK not yet implemented");
+
+#[cfg(CONFIG_DEBUG_MUTEXES)]
+compile_error!("CONFIG_DEBUG_MUTEX not yet implemented");*/
+
+pub struct Mutex<T: ?Sized> {
+    lock: bindings::mutex,
+    data: UnsafeCell<T>,
+}
+
+impl<T> Mutex<T> {
+    /// Create a new Mutex
+    pub fn new(data: T, _name: &'static str) -> Self {
+
+        let lock = bindings::mutex::default();
+        // TODO: write mutex debug traces like .magic.
+        // TODO: use name in the debug version
+
+        Self {
+            data: UnsafeCell::new(data),
+            lock
+        }
+    }
+}
+
+impl<T: ?Sized> Mutex<T> {
+    /// acquire a lock on the mutex
+    #[cfg(CONFIG_DEBUG_LOCK_ALLOC)]
+    pub fn lock<'a>(&'a self) -> MutexGuard<'a, T> {
+        unsafe {
+            //bindings::mutex_lock(&self.lock as *const bindings::mutex as *mut bindings::mutex);
+            bindings::mutex_lock_nested(&self.lock as *const bindings::mutex as *mut bindings::mutex, 0);
```

Comment:
Clarify that you need to call the actual function since the other one is a macro.

Reply:
added a comment